app/upgrades/v20/upgrades.go

```diff
@@ -152,16 +154,148 @@ func MigrateICSLegacyProposals(ctx sdk.Context, msgServer providertypes.MsgServe
+		ValidatorSetCap:                   proposal.ValidatorSetCap,
+		Allowlist:                         proposal.Allowlist,
+		Denylist:                          proposal.Denylist,
+		Authority:                         authority,
+		MinStake:                          proposal.MinStake,
+		AllowInactiveVals:                 proposal.AllowInactiveVals,
+	}
+}
+
+func ConsumerRemovalProposalToMsgConsumerRemoval(proposal providertypes.ConsumerRemovalProposal, authority string) providertypes.MsgConsumerRemoval {
+	return providertypes.MsgConsumerRemoval{
+		ChainId:   proposal.ChainId,
+		StopTime:  proposal.StopTime,
+		Authority: authority,
+	}
+}
+
+func ConsumerModificationProposalToMsgConsumerModification(proposal providertypes.ConsumerModificationProposal, authority string) providertypes.MsgConsumerModification {
+	return providertypes.MsgConsumerModification{
+		Title:              proposal.Title,
+		Description:        proposal.Description,
+		ChainId:            proposal.ChainId,
+		Top_N:              proposal.Top_N,
+		ValidatorsPowerCap: proposal.ValidatorsPowerCap,
+		ValidatorSetCap:    proposal.ValidatorSetCap,
+		Allowlist:          proposal.Allowlist,
+		Denylist:           proposal.Denylist,
+		Authority:          authority,
+		MinStake:           proposal.MinStake,
+		AllowInactiveVals:  proposal.AllowInactiveVals,
+	}
+}
+
+func ChangeRewardDenomsProposalToMsgChangeRewardDenoms(proposal providertypes.ChangeRewardDenomsProposal, authority string) providertypes.MsgChangeRewardDenoms {
+	return providertypes.MsgChangeRewardDenoms{
+		DenomsToAdd:    proposal.DenomsToAdd,
+		DenomsToRemove: proposal.DenomsToRemove,
+		Authority:      authority,
+	}
+}
+
+func MigrateICSProposal(
+	ctx sdk.Context,
+	msgServer providertypes.MsgServer,
+	providerKeeper providerkeeper.Keeper,
+	govKeeper govkeeper.Keeper,
+	proposal govtypes.Proposal) error {
+	// ignore proposals that were rejected or failed
+	if proposal.Status != govtypes.StatusDepositPeriod &&
+		proposal.Status != govtypes.StatusVotingPeriod &&
+		proposal.Status != govtypes.StatusPassed {
+		return nil
+	}
+
+	proposalMessages := proposal.GetMessages()
+	for index, proposalMsg := range proposalMessages {
+		switch msg := proposalMsg.GetCachedValue().(type) {
+		case *providertypes.MsgConsumerAddition:
+			err := MigrateMsgConsumerAddition(
+				ctx,
+				msgServer,
+				providerKeeper,
+				govKeeper,
+				proposal.Id,
+				*msg,
+				index,
+			)
+			if err != nil {
+				return err
+			}
+		case *providertypes.MsgConsumerRemoval:
+			err := MigrateMsgConsumerRemoval(
+				ctx,
+				msgServer,
+				providerKeeper,
+				govKeeper,
+				proposal.Id,
+				*msg,
+				index,
+			)
+			if err != nil {
+				return err
+			}
+		case *providertypes.MsgConsumerModification:
+			err := MigrateMsgConsumerModification(
+				ctx,
+				providerKeeper,
+				govKeeper,
+				proposal.Id,
+				*msg,
+				index,
+			)
+			if err != nil {
+				return err
+			}
+		case *providertypes.MsgChangeRewardDenoms:
```

Comment:
This doesn't need to be migrated. This message was not changed.

Reply:
Good point. Removed.